ibis/expr/operations.py

```diff
@@ -12,25 +12,154 @@
-
+from functools import partial
 from ibis.expr.schema import HasSchema, Schema
-from ibis.expr.rules import value, string, number, integer, boolean, list_of
-from ibis.expr.types import (Node, as_value_expr, Expr,
-                             ValueExpr, ColumnExpr, TableExpr,
-                             ValueOp, _safe_repr, distinct_roots)
-import ibis.common as com
-import ibis.expr.datatypes as dt
-import ibis.expr.rules as rules
-import ibis.expr.types as ir
+
 import ibis.util as util
+import ibis.common as com
 import ibis.compat as compat
+import ibis.expr.types as ir
+import ibis.expr.rules as rlz
+import ibis.expr.schema as sch
+import ibis.expr.datatypes as dt
+from ibis.expr.signature import Annotable, Argument as Arg
+
+
+def _safe_repr(x, memo=None):
+    return x._repr(memo=memo) if isinstance(x, (ir.Expr, Node)) else repr(x)
+
+
+# TODO: move to analysis
+def distinct_roots(*expressions):
+    roots = toolz.concat(
+        expression._root_tables() for expression in expressions
+    )
+    return list(toolz.unique(roots, key=id))
+
+
+class Node(Annotable):
+
+    __slots__ = ('_expr_cached',)
+
+    def __repr__(self):
+        return self._repr()
+
+    def _repr(self, memo=None):
+        if memo is None:
+            from ibis.expr.format import FormatMemo
+            memo = FormatMemo()
+
+        opname = type(self).__name__
+        pprint_args = []
+
+        def _pp(x):
+            return _safe_repr(x, memo=memo)
+
+        for x in self.args:
+            if isinstance(x, (tuple, list)):
+                pp = repr([_pp(y) for y in x])
+            else:
+                pp = _pp(x)
+            pprint_args.append(pp)
+
+        return '%s(%s)' % (opname, ', '.join(pprint_args))
+
+    def blocks(self):
+        # The contents of this node at referentially distinct and may not be
+        # analyzed deeper
+        return False
+
+    def flat_args(self):
+        for arg in self.args:
+            if not isinstance(arg, six.string_types) and isinstance(
+                arg, collections.Iterable
+            ):
+                for x in arg:
+                    yield x
+            else:
+                yield arg
+
+    def equals(self, other, cache=None):
+        if cache is None:
+            cache = {}
+
+        if (self, other) in cache:
+            return cache[(self, other)]
+
+        if self is other:
+            cache[(self, other)] = True
+            return True
+
+        if type(self) != type(other):
+            cache[(self, other)] = False
+            return False
+
+        if len(self.args) != len(other.args):
+            cache[(self, other)] = False
+            return False
+
+        for left, right in zip(self.args, other.args):
+            if not all_equal(left, right, cache=cache):
+                cache[(self, other)] = False
+                return False
+        cache[(self, other)] = True
+        return True
+
+    def is_ancestor(self, other):
```

Comment:
Is this being used anywhere? It seems like it's really not that different from just a straight-up call to `equals`.